articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+
+## Passing data to the callee and back
+
+Passing data is done in two different ways. The first is if you compile a program for 32 bits (x86) then all arguments are passed on the stack. The second way is if you compile for 64 bits (x64 architecture) then registers are used to pass arguments. If you have a lot of arguments then some are put onto the stack.
+
+The reason we can use the stack to pass arguments is because the stack is predictable enough that at compile-time, offsets from the top of the stack can be used. Here is a picture to make the point more clear.
+
+![alt text](image1.png "Call stack")
+
+*`rsp` and `esp` refer to the same register. `esp` is used in an x86 architecture and is the lower 4 bytes of `rsp` which is used in an x64 architecture.*
+*Note: a lot of the time the stack is drawn upside down for whatever reason so make sure you look at the direction of increasing addresses.*
+
+The stack pointer, `rsp` (`esp` in x86), always points to the top of the stack. From there we can apply an offset known at compile time to get the arguments. This offset is calculated based on the size that the callee allocated for local variables, the numbers of callee-saved registers saved, one pointer sized value for the return address, and finally the parameter sizes and alignments or each parameter before it. Note that when a function allocates space for local variable it will allocate it insuring there is a [16-byte alinement](https://stackoverflow.com/a/49397524/9664285) (some times even 32-bytes or 64-bytes depending on SIMD hardware being used <!-- well right now more then 16 is never required --> but you can't rely on that) by adding padding so that everything that follows has a known alignment.
+
+So, let's look at how parameters are passed to the callee in C when there are registers available to use. Note, this is only the case with the architecture is x64. We figure out how each argument is passed individually based on its type-kind. We do this for each argument:
+
+- If the type-kind of the argument is an Integers then we will pick the next available general-purpose argument register (they are in order `rdi`, `rsi`, `rdx`, `rcx`, `r8`, `r9`). If there are no available registers then the argument will be placed in the next argument position on the stack.
+
+- If the type-kind of the argument is a Float then we will pick the next available float/vector argument register (they are `xmm0` to `xmm7`). If there are no available registers then the argument will be placed in the next argument position on the stack.
+
+- If the type-kind is a Pointer then most of the time the same steps as Integers type-kinds will be used. If there are bounds associated with the pointer then more steps are taken, however, that is outside the scope of this article (read [this](https://software.intel.com/sites/default/files/article/402129/mpx-linux64-abi.pdf#section.3.2) page 24 for more information).

```

Comment:
the same steps as Integers type-kinds will be used. -> we can follow the same steps as Integer type-kinds.

Reply:
steps are taken, however, -> steps are needed, however,